src/connector/src/sink/snowflake.rs

```diff
@@ -236,32 +228,105 @@ impl SnowflakeSinkWriter {
             epoch: 0,
-        }
+            // will be (lazily) initialized after the begin of each epoch
+            // when some data is ready to be upload
+            streaming_uploader: None,
+        })
+    }
+
+    /// update the streaming uploader as well as the file suffix.
+    /// note: should *only* be called iff after a new epoch begins,
+    /// and `streaming_upload` being called the first time.
+    /// i.e., lazily initialization of the internal `streaming_uploader`.
+    async fn new_streaming_uploader(&mut self) -> Result<()> {
+        let file_suffix = self.file_suffix();
+        let path = generate_s3_file_name(self.s3_client.s3_path(), &file_suffix);
+        let uploader = self
+            .s3_client
+            .opendal_s3_engine
+            .streaming_upload(&path)
+            .await
+            .map_err(|err| {
+                SinkError::Snowflake(format!(
+                    "failed to create the streaming uploader of opendal s3 engine for epoch {}, error: {}",
+                    self.epoch,
+                    err
+                ))
+            })?;
+        self.streaming_uploader = Some((uploader, file_suffix));
+        Ok(())
+    }
+
+    /// should be *only* called after `commit` per epoch.
+    fn reset_streaming_uploader(&mut self) {
+        self.streaming_uploader = None;
     }
 
-    /// reset the `payload` and `row_counter`.
-    /// shall *only* be called after a successful sink to s3.
-    fn reset(&mut self) {
-        self.payload.clear();
-        self.row_counter = 0;
+    /// the `Option` is the flag to determine if there is data to upload.
+    fn has_data(&self) -> bool {
+        self.streaming_uploader.is_some()
     }
 
-    fn at_sink_threshold(&self) -> bool {
-        self.row_counter >= self.max_batch_row_num
+    /// write data to the current streaming uploader for this epoch.
+    async fn streaming_upload(&mut self, data: Bytes) -> Result<()> {
+        if !self.has_data() {
+            // lazily initialization
+            self.new_streaming_uploader().await?;
+        }
+        let Some(uploader) = self.streaming_uploader.as_mut() else {
+            return Err(SinkError::Snowflake(format!(
+                "expect streaming uploader to be properly initialized when performing streaming upload for epoch {}",
+                self.epoch
+                ))
+            );
+        };
+        uploader
+            .0
+            .write_bytes(data)
+            .await
+            .map_err(|err| {
+                SinkError::Snowflake(format!(
+                    "failed to write bytes when streaming uploading to s3 for snowflake sink, error: {}",
+                    err
+                ))
+            })?;
+        Ok(())
+    }
+
+    /// finalize streaming upload for this epoch.
+    /// ensure all the data has been properly uploaded to intermediate s3.
+    async fn finish_streaming_upload(&mut self) -> Result<String> {
```

Comment:
A better design will be changing the return type to `Result<Option<String>>`. When the `streaming_uploader` is `None` (indicating there is no new data), we simply return `None`. In the `commit` method we only commit to snowflake when this method returns `Some(file_name)`.